### PR TITLE
python-common: only validate host_pattern if present

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

Fixes: https://tracker.ceph.com/issues/45203

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by smithfarm** [comment]
Note: @sebastian-philipp appears to have changed the scope of the tracker https://tracker.ceph.com/issues/45203 ... I guess this PR just fixes the originally reported bug, but doesn't implement the feature?

File: src/python-common/ceph/deployment/drive_group.py [hunk 1]
```diff
@@ -241,7 +241,8 @@ def validate(self):
         # type: () -> None
         super(DriveGroupSpec, self).validate()
 
-        if not isinstance(self.placement.host_pattern, six.string_types):
+        if not isinstance(self.placement.host_pattern, six.string_types) and \
+                self.placement.host_pattern is not None:
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/drive_group.py:245]: Yes, we also need to adapt

https://github.com/ceph/ceph/blob/fee0cda09237ea6d8d9c8cfec31cd3f3555d7dfa/src/pybind/mgr/cephadm/module.py#L2102

Comment (jschmid1) [reply src/python-common/ceph/deployment/drive_group.py:245]: ack

**Comment by sebastian-philipp** [comment]
ping?

**Comment by jschmid1** [comment]
I noticed that we don't have a uniform`get_hosts` return type across our orchestrators.

While we have:

**cephadm**:

* self.cache.get_hosts() -> List[str] 
goes via `HostCache` and looks for `daemons` hostnames

* self._get_hosts() -> List[str]
goes via inventory

* self.get_hosts() -> List[orchestrator.HostSpec] 
gets information from `list(inventory.all_specs())` very odd..

* several instances where we get hosts via one of the above funcs and shove the return in `HostPlacementSpec` 

**rook**:

* self.get_hosts -> List[orchestrator.HostSpec]
gets this from `rook_cluster.get_node_names()`

**test_orchestrator**:

* self.get_hosts -> List[orchestrator.HostSpec]
via the _inventory object


This probably needs to be cleaned up some day. I'd argue that this is out of scope of _this_ patchset.

**Comment by sebastian-philipp** [comment]
Plus HostPlacementSpec is odd:

https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/module.py#L1181-L1186

to get the hosts matching a label, you need to call 

    self._get_hosts(spec.placement.label)

to get the host matching a patterm, you need to call

    spec.placement.pattern_matches_hosts(self.inventory.keys())

Would be great to have a unified 

    HostPlacementSpec.get_matching_hosts(hosts: List[HostSpec])

File: src/python-common/ceph/deployment/service_spec.py [hunk 2]
```diff
@@ -173,11 +173,23 @@ def set_hosts(self, hosts):
         # in the orchestrator backend.
         self.hosts = hosts
 
-    def pattern_matches_hosts(self, all_hosts):
-        # type: (List[str]) -> List[str]
-        if not self.host_pattern:
+    def filter_matching_hosts(self, _get_hosts_func: Callable) -> List[str]:
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/service_spec.py:176]: What about simplifying this to
```suggestion
    def filter_matching_hosts(self, hosts: List[HostSpec]) -> List[str]:
```
? Is there a need to encapsulate this in a function?

File: src/pybind/mgr/rook/module.py [hunk 3]
```diff
@@ -492,10 +492,9 @@ def create_osds(self, drive_group):
         if drive_group.data_directories:
             targets += drive_group.data_directories
 
-        def execute(all_hosts_):
-            # type: (List[orchestrator.HostSpec]) -> orchestrator.Completion
-            all_hosts = [h.hostname for h in all_hosts_]
-            matching_hosts = drive_group.placement.pattern_matches_hosts(all_hosts)
+        def execute():
+            # type: () -> orchestrator.Completion
+            matching_hosts = drive_group.placement.filter_matching_hosts(self.get_hosts)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py:497]: For this, you need to split `self.get_hosts` into to methods:

```python
class RookOrchestrator:
    def get_hosts(self) -> orchestrator.Completion:
        ...

    def _get_hosts(self) -> List[HostSpec]
        ...
```

then you can do

```suggestion
            matching_hosts = drive_group.placement.filter_matching_hosts(self._get_hosts)
```

File: src/pybind/mgr/test_orchestrator/module.py [hunk 4]
```diff
@@ -259,7 +259,7 @@ def run(all_hosts):
             # type: (List[orchestrator.HostSpec]) -> None
             drive_group.validate()
             if drive_group.placement.host_pattern:
-                if not drive_group.placement.pattern_matches_hosts([h.hostname for h in all_hosts]):
+                if not drive_group.placement.filter_matching_hosts(self.get_hosts):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/test_orchestrator/module.py:262]: `self.get_hosts` returns a completion, not a `List[HostSpec]`

**Comment by sebastian-philipp** [comment]
> I noticed that we don't have a uniform`get_hosts` return type across our orchestrators.
> 
> While we have:
> 
> **cephadm**:
> 
> * self.cache.get_hosts() -> List[str]
>   goes via `HostCache` and looks for `daemons` hostnames
> * self._get_hosts() -> List[str]
>   goes via inventory
> * self.get_hosts() -> List[orchestrator.HostSpec]

`self.get_hosts() -> Completion`

>   gets information from `list(inventory.all_specs())` very odd..
> * several instances where we get hosts via one of the above funcs and shove the return in `HostPlacementSpec`
> 
> **rook**:
> 
> * self.get_hosts -> List[orchestrator.HostSpec]

`self.get_hosts() -> Completion`

>   gets this from `rook_cluster.get_node_names()`
> 
> **test_orchestrator**:
> 
> * self.get_hosts -> List[orchestrator.HostSpec]

`self.get_hosts() -> Completion`

>   via the _inventory object
> 
> This probably needs to be cleaned up some day. I'd argue that this is out of scope of _this_ patchset.

**Comment by sebastian-philipp** [comment]
```
=================================== FAILURES ===================================
_________________________ test_DriveGroup[test_input0] _________________________

test_input = [{'data_devices': {'paths': ['/dev/sda']}, 'placement': {'host_pattern': 'hostname'}, 'service_id': 'testing_drivegroup', 'service_type': 'osd'}]

    @pytest.mark.parametrize("test_input",
    [
        (
            [  # new style json
                {
                    'service_type': 'osd',
                    'service_id': 'testing_drivegroup',
                    'placement': {'host_pattern': 'hostname'},
                    'data_devices': {'paths': ['/dev/sda']}
                }
            ]
        ),
    ])
    def test_DriveGroup(test_input):
        dg = [DriveGroupSpec.from_json(inp) for inp in test_input][0]
>       assert dg.placement.pattern_matches_hosts(['hostname']) == ['hostname']
E       AttributeError: 'PlacementSpec' object has no attribute 'pattern_matches_hosts'

ceph/tests/test_drive_group.py:27: AttributeError
___________________________ test_drivegroup_pattern ____________________________

    def test_drivegroup_pattern():
        dg = DriveGroupSpec(PlacementSpec(host_pattern='node[1-3]'), data_devices=DeviceSelection(all=True))
>       assert dg.placement.pattern_matches_hosts(['node{}'.format(i) for i in range(10)]) == ['node1', 'node2', 'node3']
E       AttributeError: 'PlacementSpec' object has no attribute 'pattern_matches_hosts'

ceph/tests/test_drive_group.py:56: AttributeError
```

File: src/python-common/ceph/deployment/service_spec.py (hunk 2 above)
Comment (jschmid1) [reply src/python-common/ceph/deployment/service_spec.py:176]: Yes, this reduces the amount of calls to `get_hosts` as we can pass the arguments to it _inside_ `filter_matching_hosts()`. If we pass a hostlist to the function, we have to construct it before based on the presence of `labels`, `host_pattern` etc..

File: src/pybind/mgr/rook/module.py (hunk 3 above)
Comment (jschmid1) [reply src/pybind/mgr/rook/module.py:497]: Isn't the method I'm using already returning `List[HostSpec]` -> https://github.com/ceph/ceph/blob/9ea2ec7472bb8a580cf1bc4823298364716d26b5/src/pybind/mgr/rook/module.py#L253-L256 ?

Comment (jschmid1) [reply src/pybind/mgr/rook/module.py:497]: Ah, `deferred_read` ..  the type annotation is just wrong here. Fixing

File: src/pybind/mgr/test_orchestrator/module.py (hunk 4 above)
Comment (jschmid1) [reply src/pybind/mgr/test_orchestrator/module.py]: done

File: src/pybind/mgr/rook/module.py (hunk 3 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/module.py]: note that mypy and pycharm use different mechanics for type annotations:

consider
```
def to_int(f: Callable[[Any], str]) -> int:
    def wrapper(arg: Any) -> int:
        return f(arg)
    return wrapper

@to_int
def works_in_mypy(arg: Any) -> str:
    return str(arg)

@to_int
def works_in_pycharm(arg: Any) -> int:
    return str(arg)
```

I'm using the mypy style till now in the code base.

File: src/pybind/mgr/rook/module.py [hunk 5]
```diff
@@ -252,7 +252,6 @@ def get_inventory(self, host_filter=None, refresh=False):
 
     @deferred_read
     def get_hosts(self):
-        # type: () -> List[orchestrator.HostSpec]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py:255]: revert this?

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-05-20_10:26:42-rados-wip-swagner-testing-2020-05-20-0959-distro-basic-smithi/5072686/

```
2020-05-20T10:45:18.329 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]: Traceback (most recent call last):
2020-05-20T10:45:18.329 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 274, in _finalize
2020-05-20T10:45:18.329 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     next_result = self._on_complete(self._value)
2020-05-20T10:45:18.329 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 140, in <lambda>
2020-05-20T10:45:18.330 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     return CephadmCompletion(on_complete=lambda _: f(*args, **kwargs))
2020-05-20T10:45:18.330 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 1334, in remove_service
2020-05-20T10:45:18.330 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     self._trigger_preview_refresh(service_name=service_name)
2020-05-20T10:45:18.330 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 1455, in _trigger_preview_refresh
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     refresh_hosts = self.resolve_hosts_for_osdspecs(specs=specs, service_name=service_name)
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 1441, in resolve_hosts_for_osdspecs
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     return sum([spec.placement.pattern_matches_hosts(self.cache.get_hosts()) for spec in osdspecs], [])
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 1441, in <listcomp>
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]:     return sum([spec.placement.pattern_matches_hosts(self.cache.get_hosts()) for spec in osdspecs], [])
2020-05-20T10:45:18.331 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]: AttributeError: 'PlacementSpec' object has no attribute 'pattern_matches_hosts'
2020-05-20T10:45:18.332 INFO:ceph.mgr.a.smithi169.stdout:May 20 10:45:18 smithi169 bash[37783]: debug 2020-05-20T10:45:18.054+0000 7fb87ae97700 -1 mgr.server reply reply (22) Invalid argument 'PlacementSpec' object has no attribute 'pattern_matches_hosts'
```

File: src/pybind/mgr/rook/module.py (hunk 5 above)
Comment (jschmid1) [reply src/pybind/mgr/rook/module.py:255]: that would overshadow the returntype of the decorator, wouldn't it?

**Comment by jschmid1** [comment]
about to schedule another teuthology run for test_orchestrator ..

File: src/pybind/mgr/rook/module.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/rook/module.py:255]: This defines the type of the get_hosts without the decorator. The decorator is then applied on top of it.

Comment (jschmid1) [reply src/pybind/mgr/rook/module.py]: got it, done

**Comment by sebastian-philipp** [comment]
http://pulpito.ceph.com/swagner-2020-06-02_12:46:10-rados:cephadm-wip-swagner3-testing-2020-06-02-1145-distro-basic-smithi/

**Comment by jschmid1** [comment]
> http://pulpito.ceph.com/swagner-2020-06-02_12:46:10-rados:cephadm-wip-swagner3-testing-2020-06-02-1145-distro-basic-smithi/

looks good :thumbs:

**Pull request merged**